File: cpp/src/multipass.cpp
#include "parser/ast.hpp"
#include "parser/lexer.hpp"
#include "parser/parser.hpp"

#include <algorithm>
#include <bit>
#include <cstddef>
#include <cstdint>
#include <stdexcept>
#include <string>
#include <vector>

namespace mp {
namespace {

// Divide-and-conquer parser. Full complexity analysis in multipass_arena.cpp.
//
// This variant (pointer-AST) adds two improvements over the previous version:
//
//   #8  Iterator passing (ported from multipass-arena): when parseRange splits
//       at candidate k, the two sub-ranges' candidate spans are already known —
//       the left half gets [cbeg, splitIt) and the right gets [splitIt+1, cend).
//       Flat-chain sub-ranges between chain operators have ZERO depth-d
//       candidates and receive (cend, cend). The only call that still needs a
//       binary search is a paren strip (depth change). Before this fix every
//       recursive call did a binary search: O(log n) per call, O(n log² n)
//       total. Now it is O(n log n) like the arena variant.
//
//   #9  AVX2 SIMD window (same design as multipass-arena): a per-depth int8_t
//       prec array lets a single 32-byte load answer both the flat check and the
//       split decision branchlessly. Compile-time gated on __x86_64__; runtime-
//       dispatched so the same binary runs on non-AVX2 machines. Set
//       MP_NO_SIMD=1 to force the scalar path for A/B comparison.

int binPrec(TokenType t) {
    switch (t) {
        case TokenType::Plus:
        case TokenType::Minus: return 1;
        case TokenType::Star:
        case TokenType::Slash: return 2;
        case TokenType::Caret: return 4;
        default:               return -1;
    }
}
constexpr int kUnaryPrec = 3;

struct Candidate {
    std::size_t pos;
    int         prec;
    bool        rightAssoc;
    bool        unary;
};

struct Buckets {
    std::vector<std::size_t> p1, p2, caret, un;
};

constexpr std::ptrdiff_t kScanBudget = 16;
constexpr std::size_t    kNone       = static_cast<std::size_t>(-1);

std::size_t rightmostIn(const std::vector<std::size_t>& v,
                        std::size_t lo, std::size_t hi) {
    auto it = std::lower_bound(v.begin(), v.end(), hi);
    if (it == v.begin()) return kNone;
    --it;
    return (*it >= lo) ? *it : kNone;
}

bool anyIn(const std::vector<std::size_t>& v, std::size_t lo, std::size_t hi) {
    auto it = std::lower_bound(v.begin(), v.end(), lo);
    return it != v.end() && *it < hi;
}

// ── AVX2 SIMD window (runtime-dispatched) ──────────────────────────────────
#if defined(__x86_64__)
#define MP_SIMD 1
#include <cstdlib>
#include <immintrin.h>

const bool kHasSIMD = __builtin_cpu_supports("avx2") &&
                      std::getenv("MP_NO_SIMD") == nullptr;

constexpr long kWinFirst   = -1;
constexpr long kWinBuckets = -2;

__attribute__((target("avx2")))
long simdWindowSplit(const int8_t* prec, long ib, long ie) {
    const long k    = ie - ib;
    const long base = (k >= 32) ? ie - 32 : ib;
    const uint32_t valid = (k >= 32) ? 0xFFFFFFFFu : ((1u << k) - 1);
    const __m256i v = _mm256_loadu_si256(
        reinterpret_cast<const __m256i*>(prec + base));
    const auto m1 = static_cast<uint32_t>(_mm256_movemask_epi8(
        _mm256_cmpeq_epi8(v, _mm256_set1_epi8(1)))) & valid;
    if (m1) return base + 31 - std::countl_zero(m1);
    if (k > 32) return kWinBuckets;
    const auto m2 = static_cast<uint32_t>(_mm256_movemask_epi8(
        _mm256_cmpeq_epi8(v, _mm256_set1_epi8(2)))) & valid;
    if (m2) return base + 31 - std::countl_zero(m2);
    return kWinFirst;
}

__attribute__((target("avx2")))
int simdWindowFlat(const int8_t* prec, long ib, long ie) {
    const long k = ie - ib;
    const int8_t p0 = prec[ib];
    if (p0 == 3) return 0;
    const long w = (k < 32) ? k : 32;
    const uint32_t valid = (w == 32) ? 0xFFFFFFFFu : ((1u << w) - 1);
    const __m256i v = _mm256_loadu_si256(
        reinterpret_cast<const __m256i*>(prec + ib));
    const auto eq = static_cast<uint32_t>(_mm256_movemask_epi8(
        _mm256_cmpeq_epi8(v, _mm256_set1_epi8(p0))));
    if ((eq & valid) != valid) return 0;
    return (k <= 32) ? 1 : -1;
}
#endif  // __x86_64__

class MultiPass final : public IParser {
    using CIt = std::vector<Candidate>::const_iterator;

public:
    const char* name() const override { return "multipass"; }

    ExprPtr parse(std::string_view src) override {
        tokens_ = tokenize(src);
        buildCandidates();
        const std::size_t n = tokens_.size() - 1;
        auto [cbeg, cend] = candRange(0, n, 0);
        return parseRange(0, n, 0, cbeg, cend);
    }

private:
    std::vector<Token>                  tokens_;
    std::vector<std::vector<Candidate>> candsByDepth_;
    std::vector<Buckets>                bucketsByDepth_;
#if MP_SIMD
    std::vector<std::vector<int8_t>>    precByDepth_;
#endif
    std::vector<std::size_t>            parenMatch_;

    void buildCandidates() {
        for (auto& v : candsByDepth_) v.clear();
        for (auto& b : bucketsByDepth_) {
            b.p1.clear(); b.p2.clear(); b.caret.clear(); b.un.clear();
        }
#if MP_SIMD
        for (auto& p : precByDepth_) p.clear();
#endif
        const std::size_t n = tokens_.size();
        parenMatch_.assign(n, 0);
        int  depth = 0;
        bool expectOperand = true;
        std::vector<std::size_t> stack;

        for (std::size_t i = 0; i < n - 1; ++i) {
            switch (tokens_[i].type) {
                case TokenType::LParen:
                    stack.push_back(i);
                    ++depth; expectOperand = true;
                    break;
                case TokenType::RParen:
                    if (!stack.empty()) {
                        const std::size_t open = stack.back(); stack.pop_back();
                        parenMatch_[open] = i;
                        parenMatch_[i]    = open;
                    }
                    --depth; expectOperand = false;
                    break;
                case TokenType::Number:
                case TokenType::Ident: expectOperand = false; break;
                case TokenType::Plus:
                case TokenType::Minus:
                case TokenType::Star:
                case TokenType::Slash:
                case TokenType::Caret: {
                    Candidate c;
                    c.pos = i;
                    if (expectOperand) {
                        if (tokens_[i].type != TokenType::Plus &&
                            tokens_[i].type != TokenType::Minus)
                            throw std::runtime_error("unexpected operator");
                        c.prec = kUnaryPrec; c.rightAssoc = false; c.unary = true;
                    } else {
                        c.prec       = binPrec(tokens_[i].type);
                        c.rightAssoc = (tokens_[i].type == TokenType::Caret);
                        c.unary      = false;
                    }
                    const std::size_t d = static_cast<std::size_t>(depth);
                    if (depth >= static_cast<int>(candsByDepth_.size())) {
                        candsByDepth_.resize(d + 1);
                        bucketsByDepth_.resize(d + 1);
#if MP_SIMD
                        precByDepth_.resize(d + 1);
#endif
                    }
                    candsByDepth_[d].push_back(c);
#if MP_SIMD
                    precByDepth_[d].push_back(static_cast<int8_t>(c.prec));
#endif
                    auto& bk = bucketsByDepth_[d];
                    if (c.unary)          bk.un.push_back(i);
                    else if (c.prec == 1) bk.p1.push_back(i);
                    else if (c.prec == 2) bk.p2.push_back(i);
                    else                  bk.caret.push_back(i);
                    expectOperand = true;
                    break;
                }
                default: break;
            }
        }
#if MP_SIMD
        for (auto& p : precByDepth_) p.insert(p.end(), 32, 0);
#endif
    }

    struct ByPos {
        bool operator()(const Candidate& c, std::size_t v) const { return c.pos < v; }
    };

    // O(log n) — only called on the initial parse and paren strips.
    std::pair<CIt, CIt> candRange(std::size_t lo, std::size_t hi, int depth) const {
        static const std::vector<Candidate> empty;
        const auto& v = (depth >= 0 && depth < static_cast<int>(candsByDepth_.size()))
                        ? candsByDepth_[static_cast<std::size_t>(depth)] : empty;
        auto b = std::lower_bound(v.begin(), v.end(), lo, ByPos{});
        auto e = std::lower_bound(b,         v.end(), hi, ByPos{});
        return {b, e};
    }

    CIt findSplitBuckets(CIt beg, CIt end, int depth) const {
        const auto& bk = bucketsByDepth_[static_cast<std::size_t>(depth)];
        const std::size_t blo = beg->pos, bhi = end[-1].pos + 1;
        std::size_t p = rightmostIn(bk.p1, blo, bhi);
        if (p == kNone) p = rightmostIn(bk.p2, blo, bhi);
        if (p == kNone) return beg;
        return std::lower_bound(beg, end, p, ByPos{});
    }

    bool flatByBuckets(CIt beg, CIt end, int depth) const {
        const auto& bk = bucketsByDepth_[static_cast<std::size_t>(depth)];
        const std::size_t blo = beg->pos, bhi = end[-1].pos + 1;
        if (anyIn(bk.un, blo, bhi)) return false;
        const int classes = (anyIn(bk.p1, blo, bhi) ? 1 : 0) +
                            (anyIn(bk.p2, blo, bhi) ? 1 : 0) +
                            (anyIn(bk.caret, blo, bhi) ? 1 : 0);
        return classes == 1;
    }

    CIt findSplit(CIt beg, CIt end, std::size_t lo, int depth) const {
        if (beg == end) return end;
        CIt best = end;
        const CIt stop = (end - beg > kScanBudget) ? end - kScanBudget : beg;
        for (auto it = end; it != stop; ) {
            --it;
            if (it->unary && it->pos != lo) continue;
            if (best == end || it->prec < best->prec) {
                best = it;
                if (best->prec == 1 && !best->rightAssoc) return best;
            } else if (it->prec == best->prec && best->rightAssoc) {
                best = it;
            }
        }
        if (stop == beg) return best;
        return findSplitBuckets(beg, end, depth);
    }

    // cbeg/cend: already-located candidates at `depth` in [lo, hi).
    // Splits pass sliced iterators — O(1), no binary search.
    // Flat-fold sub-ranges and paren contents get (cend, cend) or a new search.
    ExprPtr parseRange(std::size_t lo, std::size_t hi, int depth,
                       CIt cbeg, CIt cend) {
        if (lo >= hi) throw std::runtime_error("empty sub-expression");

        if (cbeg != cend) {
            const int  chainPrec = cbeg->prec;
            const bool chainRa   = cbeg->rightAssoc;

            bool flat;
#if MP_SIMD
            if (kHasSIMD) {
                const auto& dv = candsByDepth_[static_cast<std::size_t>(depth)];
                const int f = simdWindowFlat(
                    precByDepth_[static_cast<std::size_t>(depth)].data(),
                    cbeg - dv.begin(), cend - dv.begin());
                flat = (f < 0) ? flatByBuckets(cbeg, cend, depth) : (f != 0);
            } else
#endif
            {
                flat = true;
                const CIt scanEnd =
                    (cend - cbeg > kScanBudget) ? cbeg + kScanBudget : cend;
                for (auto it = cbeg; it != scanEnd; ++it)
                    if (it->unary || it->prec != chainPrec) { flat = false; break; }
                if (flat && scanEnd != cend)
                    flat = flatByBuckets(cbeg, cend, depth);
            }

            if (flat) {
                if (!chainRa) {
                    ExprPtr acc = parseRange(lo, cbeg->pos, depth, cend, cend);
                    for (auto it = cbeg; it != cend; ++it) {
                        const std::size_t nextLo = it->pos + 1;
                        const std::size_t nextHi = (it + 1 != cend) ? (it+1)->pos : hi;
                        acc = binary(tokens_[it->pos].type,
                                     std::move(acc),
                                     parseRange(nextLo, nextHi, depth, cend, cend));
                    }
                    return acc;
                } else {
                    std::vector<ExprPtr> parts;
                    parts.reserve(static_cast<std::size_t>(cend - cbeg) + 1);
                    std::size_t prevLo = lo;
                    for (auto it = cbeg; it != cend; ++it) {
                        parts.push_back(parseRange(prevLo, it->pos, depth, cend, cend));
                        prevLo = it->pos + 1;
                    }
                    parts.push_back(parseRange(prevLo, hi, depth, cend, cend));
                    ExprPtr acc = std::move(parts.back());
                    auto it = cend;
                    for (std::size_t i = parts.size() - 1; i-- > 0; ) {
                        --it;
                        acc = binary(tokens_[it->pos].type,
                                     std::move(parts[i]), std::move(acc));
                    }
                    return acc;
                }
            }
        }

        // General split — iterator passed directly, no binary search.
        CIt splitIt;
#if MP_SIMD
        if (kHasSIMD) {
            if (cbeg == cend) {
                splitIt = cend;
            } else {
                const auto& dv = candsByDepth_[static_cast<std::size_t>(depth)];
                const long r = simdWindowSplit(
                    precByDepth_[static_cast<std::size_t>(depth)].data(),
                    cbeg - dv.begin(), cend - dv.begin());
                splitIt = (r == kWinBuckets) ? findSplitBuckets(cbeg, cend, depth)
                        : (r == kWinFirst)   ? cbeg
                                             : dv.begin() + r;
            }
        } else
#endif
        splitIt = findSplit(cbeg, cend, lo, depth);

        if (splitIt != cend) {
            if (splitIt->unary)
                return unary(tokens_[splitIt->pos].type,
                             parseRange(splitIt->pos + 1, hi, depth,
                                        splitIt + 1, cend));
            return binary(tokens_[splitIt->pos].type,
                          parseRange(lo,               splitIt->pos, depth, cbeg,        splitIt),
                          parseRange(splitIt->pos + 1, hi,           depth, splitIt + 1, cend));
        }

        // No depth-d operator: paren group or leaf.
        if (tokens_[lo].type == TokenType::LParen &&
            parenMatch_[lo] == hi - 1) {
            auto [b2, e2] = candRange(lo + 1, hi - 1, depth + 1);
            return parseRange(lo + 1, hi - 1, depth + 1, b2, e2);
        }
        if (hi - lo == 1) {
            if (tokens_[lo].type == TokenType::Number) return number(tokens_[lo].value);
            if (tokens_[lo].type == TokenType::Ident)
                return variable(static_cast<int>(tokens_[lo].value));
        }
        throw std::runtime_error("syntax error at position " +
                                 std::to_string(tokens_[lo].pos));
    }
};

}  // namespace

std::unique_ptr<IParser> make_multipass() {
    return std::make_unique<MultiPass>();
}

}  // namespace mp

File: rust/src/multipass.rs
//! The top-down divide-and-conquer family ("multipass" in the original,
//! wrong-order sense): find the loosest operator of a range, split, recurse.
//! Port of python/mathparser/evaluators.py `_MP`: candidates are bucketed per
//! paren depth, a right-to-left scan finds the split with a SCAN_BUDGET, and
//! long single-level runs fold from per-precedence buckets so the two former
//! quadratic shapes (powchain / towerchain) are O(n log n). `with_sparse`
//! selects the sparse-table RMQ split (multipass-bfs).

use crate::builder::{op_of, Builder};
use crate::classics::{bin_prec, UNARY_PREC};
use crate::lexer::{Kind, Token};
use crate::Error;

#[derive(Clone, Copy)]
struct Cand {
    pos: u32,
    prec: u8,
    right_assoc: bool,
    unary: bool,
}

#[inline]
fn better(a: &Cand, b: &Cand) -> bool {
    if a.prec != b.prec {
        return a.prec < b.prec;
    }
    if a.right_assoc { a.pos < b.pos } else { a.pos > b.pos }
}

const SCAN_BUDGET: usize = 16;

/// Per-eval working state; the vectors are owned by the evaluator and reused.
#[derive(Default)]
pub struct MpState {
    cbd: Vec<Vec<Cand>>,      // candidates per depth
    nd: Vec<Vec<u32>>,        // next-class-change index per depth
    pm: Vec<u32>,             // paren match
    bks: Vec<[Vec<u32>; 2]>,  // lazily built prec-1 / prec-2 buckets per depth
    bks_built: bool,
    st: Vec<Vec<Vec<u32>>>,   // sparse table per depth (bfs only)
}

struct Mp<'t, 's> {
    t: &'t [Token],
    s: &'s mut MpState,
    with_sparse: bool,
}

pub fn mp_parse<B: Builder>(tokens: &[Token], b: &mut B, st: &mut MpState, with_sparse: bool) -> Result<B::V, Error> {
    build_candidates(tokens, st)?;
    st.bks_built = false;
    if with_sparse {
        build_sparse(st);
    }
    let mut m = Mp { t: tokens, s: st, with_sparse };
    m.range(b, 0, tokens.len() - 1, 0)
}

fn build_candidates(t: &[Token], s: &mut MpState) -> Result<(), Error> {
    for v in &mut s.cbd { v.clear(); }
    s.pm.clear();
    s.pm.resize(t.len(), 0);
    let mut depth = 0usize;
    let mut expect_operand = true;
    let mut stack: Vec<u32> = Vec::new();
    for i in 0..t.len() - 1 {
        let k = t[i].kind;
        match k {
            Kind::LParen => { stack.push(i as u32); depth += 1; expect_operand = true; }
            Kind::RParen => {
                let o = stack.pop().ok_or_else(|| Error("mismatched parenthesis".into()))?;
                s.pm[o as usize] = i as u32; s.pm[i] = o;
                depth -= 1; expect_operand = false;
            }
            Kind::Num | Kind::Ident => expect_operand = false,
            Kind::End => {}
            _ => {
                let c = if expect_operand {
                    if !matches!(k, Kind::Plus | Kind::Minus) { return Err(Error("unexpected operator".into())); }
                    Cand { pos: i as u32, prec: UNARY_PREC, right_assoc: false, unary: true }
                } else {
                    Cand { pos: i as u32, prec: bin_prec(k), right_assoc: k == Kind::Caret, unary: false }
                };
                while depth >= s.cbd.len() { s.cbd.push(Vec::new()); }
                s.cbd[depth].push(c);
                expect_operand = true;
            }
        }
    }
    // next-class-change index
    while s.nd.len() < s.cbd.len() { s.nd.push(Vec::new()); }
    for (d, vd) in s.cbd.iter().enumerate() {
        let k = vd.len();
        let nd = &mut s.nd[d];
        nd.clear();
        nd.resize(k, 0);
        for i in (0..k).rev() {
            nd[i] = if i + 1 < k && vd[i + 1].prec == vd[i].prec { nd[i + 1] } else { (i + 1) as u32 };
        }
    }
    Ok(())
}

fn build_buckets(s: &mut MpState) {
    while s.bks.len() < s.cbd.len() { s.bks.push([Vec::new(), Vec::new()]); }
    for (d, vd) in s.cbd.iter().enumerate() {
        let [b1, b2] = &mut s.bks[d];
        b1.clear(); b2.clear();
        for (i, c) in vd.iter().enumerate() {
            if c.unary { continue; }
            if c.prec == 1 { b1.push(i as u32); } else if c.prec == 2 { b2.push(i as u32); }
        }
    }
    s.bks_built = true;
}

fn build_sparse(s: &mut MpState) {
    while s.st.len() < s.cbd.len() { s.st.push(Vec::new()); }
    for (d, v) in s.cbd.iter().enumerate() {
        let n = v.len();
        let tbl = &mut s.st[d];
        tbl.clear();
        if n == 0 { continue; }
        let logn = usize::BITS as usize - n.leading_zeros() as usize;
        tbl.push((0..n as u32).collect());
        for k in 1..logn {
            let half = 1usize << (k - 1);
            let prev = &tbl[k - 1];
            let mut row = vec![0u32; n];
            for i in 0..=(n - (1usize << k)) {
                let (a, b) = (prev[i], prev[i + half]);
                row[i] = if better(&v[a as usize], &v[b as usize]) { a } else { b };
            }
            tbl.push(row);
        }
    }
}

#[inline]
fn lower_bound_pos(v: &[Cand], x: u32) -> usize {
    v.partition_point(|c| c.pos < x)
}

impl<'t, 's> Mp<'t, 's> {
    fn st_query(&self, depth: usize, lo: usize, hi: usize) -> i64 {
        if lo >= hi { return -1; }
        let v = &self.s.cbd[depth];
        let t = &self.s.st[depth];
        let k = (usize::BITS - (hi - lo).leading_zeros() - 1) as usize;
        let (a, b) = (t[k][lo], t[k][hi - (1usize << k)]);
        (if better(&v[a as usize], &v[b as usize]) { a } else { b }) as i64
    }

    /// Returns split index, -1 (no candidate: leaf/paren) or -2 (budget
    /// exhausted: fold the level from the buckets).
    fn split(&self, b: usize, e: usize, lo: usize, depth: usize) -> i64 {
        if b >= e { return -1; } // no candidates (depth may exceed the table)
        let v = &self.s.cbd[depth];
        if self.with_sparse {
            let si = self.st_query(depth, b, e);
            if si < 0 { return -1; }
            let c = v[si as usize];
            if c.unary && c.pos as usize != lo {
                for i in b..e {
                    if !v[i].unary || v[i].pos as usize == lo { return i as i64; }
                }
                return -1;
            }
            return si;
        }
        let mut best: i64 = -1;
        let stop = if e - b > SCAN_BUDGET { e - SCAN_BUDGET } else { b };
        let mut i = e;
        while i > stop {
            i -= 1;
            let c = v[i];
            if c.unary && c.pos as usize != lo { continue; }
            if best == -1 || c.prec < v[best as usize].prec {
                best = i as i64;
                if c.prec == 1 && !c.right_assoc { return best; }
            } else if c.prec == v[best as usize].prec && v[best as usize].right_assoc {
                best = i as i64;
            }
        }
        if stop == b { best } else { -2 }
    }

    fn level_fold<B: Builder>(&mut self, bl: &mut B, b: usize, e: usize, lo: usize, hi: usize, depth: usize) -> Result<B::V, Error> {
        if !self.s.bks_built { build_buckets(self.s); }
        for which in 0..2 {
            let bk = &self.s.bks[depth][which];
            let i0 = bk.partition_point(|&x| (x as usize) < b);
            let i1 = i0 + bk[i0..].partition_point(|&x| (x as usize) < e);
            if i0 < i1 {
                let idx: Vec<u32> = bk[i0..i1].to_vec();
                let v = &self.s.cbd[depth];
                let first = v[idx[0] as usize].pos as usize;
                let mut acc = self.range(bl, lo, first, depth)?;
                for j in 0..idx.len() {
                    let pos = self.s.cbd[depth][idx[j] as usize].pos as usize;
                    let nhi = if j + 1 < idx.len() { self.s.cbd[depth][idx[j + 1] as usize].pos as usize } else { hi };
                    let rhs = self.range(bl, pos + 1, nhi, depth)?;
                    acc = bl.binop(op_of(self.t[pos].kind), acc, rhs);
                }
                return Ok(acc);
            }
        }
        let c = self.s.cbd[depth][b];
        let pos = c.pos as usize;
        if c.unary {
            let operand = self.range(bl, pos + 1, hi, depth)?;
            return Ok(if self.t[pos].kind == Kind::Minus { bl.neg(operand) } else { bl.pos(operand) });
        }
        let l = self.range(bl, lo, pos, depth)?;
        let r = self.range(bl, pos + 1, hi, depth)?;
        Ok(bl.binop(op_of(self.t[pos].kind), l, r))
    }

    fn range<B: Builder>(&mut self, bl: &mut B, lo: usize, hi: usize, depth: usize) -> Result<B::V, Error> {
        if lo >= hi { return Err(Error("empty sub-expression".into())); }
        let (b, e) = if depth < self.s.cbd.len() {
            let v = &self.s.cbd[depth];
            let b = lower_bound_pos(v, lo as u32);
            (b, b + lower_bound_pos(&v[b..], hi as u32))
        } else { (0, 0) };

        // flat chain: all candidates share one precedence — fold iteratively
        if b < e {
            let c0 = self.s.cbd[depth][b];
            if !c0.unary && self.s.nd[depth][b] as usize >= e {
                if !c0.right_assoc {
                    let mut acc = self.range(bl, lo, c0.pos as usize, depth)?;
                    for i in b..e {
                        let pos = self.s.cbd[depth][i].pos as usize;
                        let nhi = if i + 1 < e { self.s.cbd[depth][i + 1].pos as usize } else { hi };
                        let rhs = self.range(bl, pos + 1, nhi, depth)?;
                        acc = bl.binop(op_of(self.t[pos].kind), acc, rhs);
                    }
                    return Ok(acc);
                }
                // right fold (^)
                let mut parts: Vec<B::V> = Vec::with_capacity(e - b + 1);
                let mut prev = lo;
                for i in b..e {
                    let pos = self.s.cbd[depth][i].pos as usize;
                    parts.push(self.range(bl, prev, pos, depth)?);
                    prev = pos + 1;
                }
                parts.push(self.range(bl, prev, hi, depth)?);
                let mut acc = parts.pop().unwrap();
                while let Some(p) = parts.pop() {
                    let pos = self.s.cbd[depth][b + parts.len()].pos as usize;
                    acc = bl.binop(op_of(self.t[pos].kind), p, acc);
                }
                return Ok(acc);
            }
        }

        let s = self.split(b, e, lo, depth);
        if s == -2 { return self.level_fold(bl, b, e, lo, hi, depth); }
        if s >= 0 {
            let c = self.s.cbd[depth][s as usize];
            let pos = c.pos as usize;
            if c.unary {
                let operand = self.range(bl, pos + 1, hi, depth)?;
                return Ok(if self.t[pos].kind == Kind::Minus { bl.neg(operand) } else { bl.pos(operand) });
            }
            let l = self.range(bl, lo, pos, depth)?;
            let r = self.range(bl, pos + 1, hi, depth)?;
            return Ok(bl.binop(op_of(self.t[pos].kind), l, r));
        }

        // paren group or leaf
        if self.t[lo].kind == Kind::LParen && self.s.pm[lo] as usize == hi - 1 {
            return self.range(bl, lo + 1, hi - 1, depth + 1);
        }
        if hi - lo == 1 {
            let tok = self.t[lo];
            if tok.kind == Kind::Num { return Ok(bl.num(tok.value)); }
            if tok.kind == Kind::Ident { return Ok(bl.var(tok.value as usize)); }
        }
        Err(Error::at("syntax error", self.t[lo].pos))
    }
}
